python/GafferSceneUI/SetQueryUI.py

```diff
@@ -0,0 +1,131 @@
+#  met:
+#
+#      * Redistributions of source code must retain the above
+#        copyright notice, this list of conditions and the following
+#        disclaimer.
+#
+#      * Redistributions in binary form must reproduce the above
+#        copyright notice, this list of conditions and the following
+#        disclaimer in the documentation and/or other materials provided with
+#        the distribution.
+#
+#      * Neither the name of John Haddon nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+from re import M
+import Gaffer
+import GafferScene
+
+Gaffer.Metadata.registerNode(
+
+	GafferScene.SetQuery,
+
+	"description",
+	"""
+	Queries the set memberships of a location, and outputs a list of
+	the sets that it belongs to.
+	""",
+
+	plugs = {
+
+		"scene" : [
+
+			"description",
+			"""
+			The scene to query.
+			""",
+
+		],
+
+		"location" : [
+
+			"description",
+			"""
+			The location to query the set memberships for.
+			""",
+
+			"plugValueWidget:type", "GafferSceneUI.ScenePathPlugValueWidget",
+			"scenePathPlugValueWidget:scene", "scene",
+			"nodule:type", "",
+
+		],
+
+		"sets" : [
+
+			"description",
+			"""
+			The sets to query.
+			""",
+
+			"nodule:type", "",
+
+		],
+
+		"inherit" : [
+
+			"description",
+			"""
+			When on, locations are treated as being in a set if an
+			ancestor location is in that set.
+			""",
+
+			"nodule:type", "",
+
+		],
+
+		"matches" : [
+
+			"description",
+			"""
+			The list of sets that the `location` is a member of. Returned in the
+			order they are listed in the `sets` plug.
+
+			> Note : When matches are inherited from a parent location, these
+			> are returned _before_ matches at this location, regardless of the
+			> order they are listed in the `sets` plug. This makes `firstMatch`
+			> consistent throughout the scene hierarchy, with children sharing
+			> the `firstMatch` of their parent.
```

Comment:
> Could you speak a bit to the motivation behind the note?

I think it's necessary for the SetSwitch. The internal Replace node will be replacing from the ancestor match down (so descendant matches are irrelevant), so SetQuery needs to tell you which set matched that ancestor, to drive the NameSwitch that chooses the replacement.

> When I try to picture a scenario with nested sets, it seems like either outcome could be desirable, and the order specified in the sets plug seems like a good way to specify which you want.

I can picture a scenario where you want the _opposite_ of what I've done, where each location's matches follow the `sets` ordering, but more local matches come first. For instance, if you wanted to put an attribute on each location matched by `"OuterSet InnerSet"` and have the "most local" attribute inheritance rules in sync with the "most local" `firstMatch` rule. You could do that with `matches[-1]` or a `lastMatch` convenience output.

But I haven't been able to picture a nested-sets scenario where you gain anything by rigid adherence to only `sets`. Did you have one in mind?

> Do you think this behaviour is more desirable, or is it just an efficiency gain, and a corner case that we don't expect to arise in practice?

There is definitely a real efficiency gain in being able to omit a set from `descendantMatches` as soon as you've included it in `matches` - particularly for the expected use cases with "root of asset" type sets, where there may only be one matching location, but huge numbers of locations below it. I suppose some of that benefit could be retained in an alternative implementation, but I _think_ there are still cases where you'd end up doing far more set matching than otherwise.

I kindof was thinking that this was mostly a corner case, and I was pleased that the SetSwitch requirement coincided with what I think is a fairly elegant and efficient implementation. But now I'm tempted to argue that this behaviour is actually more generally desirable, as it provides information about inheritance/locality that otherwise isn't available.

Reply:
I'm not sure if my examples are practical, but I feel like it's not hard to come up with them.

How about:  I want to assign an attribute for "which shader to use" to a bunch of assets, based on material sets.  I list the sets in order that I want them to override in "metal,rubber,tentacle" because some objects are both metal and tentacles, and I want them to use the metal shader.  In this case, I'm not thinking about where the set is assigned - if a set is assigned partway up the hierarchy, or at a leaf, then it shouldn't make a difference.  Maybe this is a weird way of thinking about it, but it does seem kinda natural when you're using it in a spreadsheet selector.

I guess with the current approach, I just have to get the `matches` output and then sort them myself to match the order I want ... does work, just a bit weird, and the Note in the docs was a bit unexpected for me.

Being able to omit a set from descendantMatches as soon as it's in matches should be OK as long as parent matches are being included at all, right, no matter the ordering?

It makes sense that this ordering provides extra information to people who are doing fancy scripting things ( and they can always sort to match the input if desired ).  And it makes sense that this behaviour is needed for SetSwitch.  The part I wonder most about is when using firstMatch to drive a spreadsheet ... if you're just thinking "I put this set first because I always want my spreadsheet rule to apply for that set no matter what other set something is in".  I can picture someone being confused by that, and not wanting to make them write an expression.

Reply:
> I list the sets in order that I want them to override in "metal,rubber,tentacle" because some objects are both metal and tentacles, and I want them to use the metal shader.

What does it mean for an object to be in two material sets, in this hypothetical example? In terms of shader assignments, that wouldn't be possible, but hierarchical ordering in `matches` would allow you to mimic the inherited results of a shader assignment. And if it's meaningful to be in two sets, then you can create an attribute with the two sets in it to reflect that.

> I'm not sure if my examples are practical, but I feel like it's not hard to come up with them.

I'm not sure either - it seems a bit contrived to say something could have two materials, when that isn't really true in practice. Since hierarchical ordering doesn't preclude this hypothetical use case (just makes it a pain), but is _essential_ for less hypothetical use cases, it still seems the better choice to me.

> Being able to omit a set from descendantMatches as soon as it's in matches should be OK as long as parent matches are being included at all, right, no matter the ordering?

Yes, I'm not sure what I was thinking before. But it does seem like this would require a fiddlier implementation, since the ancestor matches aren't a simple prefix, so you need to insert the matches from this location into them in the right order.

> The part I wonder most about is when using firstMatch to drive a spreadsheet ... if you're just thinking "I put this set first because I always want my spreadsheet rule to apply for that set no matter what other set something is in". I can picture someone being confused by that, and not wanting to make them write an expression.

For the Spreadsheet case, we need to consider what the results of the Spreadsheet are being used for. It seems inevitable that they're being used to provide plug values for a per-location operation of some sort, such as an AttributeProcessor or ObjectProcessor (otherwise why would the selector be varying per-location?). In the case of non-hierarchical set memberships, we have a natural user-controlled ordering from the rows. In the case of hierarchical memberships, I just can't picture wanting value resolution not to take the hierarchy into account.

Reply:
Thinking about this as "a natural user-controlled ordering from the rows" on a spreadsheet makes me feel even more like it could be confusing to sometimes not respect that order in the presence of hierarchy.

I'll try a little harder to come up with a vaguely practical sounding example.

I'm using a spreadsheet to set per location renderer attributes for some hypothetical renderer.  I have a set called FX_BREAK for any geo that has been replaced with shattered geo, and a set called VEGETATION for all plants and trees.  I add a spreadsheet row that turns off subsurface for all geo in FX_BREAK, which does what I expect.  I then add a row that sets 4 SSS samples on all geo in VEGETATION.  I am then surprised when some of the geo in FX_BREAK no longer gets subsurface turned off, because it is also vegetation ( or not, depending on hierarchy, but I've tried to come up with an example where it could make sense that VEGETATION could be defined further up the hierarchy, but FX_BREAK defined at leaf locations ).

Reply:
At first I convinced myself that this worked as long as you use `lastMatch` as the selector, so that the most local set membership is driving things. I've attached an example below (using an expression to pick the last match since we don't have that output yet). I used CustomAttributes for simplicity of inspection, but the same would work for ShaderTweaks.

- `/vegetation` is in the VEGETATION set, so we set the SSS samples, and attribute inheritance takes care of any vegetation below it.
- `/vegetation/shatteredTree` is in the VEGETATION set through inheritance, but the most local set is FX_BREAK, so that wins and we turn off SSS.
- `/vegetation/shatteredTreeBothSets` is in both sets, but FX_BREAK wins because it is matched last (via the inheritance rules), so SSS is turned off.
- `/shatteredRock` is only in the FX_BREAK set, so it gets SSS turned off.

Where it falls down is `/shatteredTreeBothSets`. Here the last match is VEGETATION (because it is a member at the leaf level, not through inheritance), and we don't turn off SSS. This _would_ be fixed by strictly following the order defined by the `SetQuery.sets` plug.

<details>

```
import Gaffer
import GafferScene
import IECore
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 3, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["Tree"] = GafferScene.Cube( "Tree" )
parent.addChild( __children["Tree"] )
__children["Tree"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"] = GafferScene.Group( "Group" )
parent.addChild( __children["Group"] )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"]["in"].addChild( GafferScene.ScenePlug( "in3", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Group"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShatteredTree"] = GafferScene.Cube( "ShatteredTree" )
parent.addChild( __children["ShatteredTree"] )
__children["ShatteredTree"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Set"] = GafferScene.Set( "Set" )
parent.addChild( __children["Set"] )
__children["Set"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["PathFilter"] = GafferScene.PathFilter( "PathFilter" )
parent.addChild( __children["PathFilter"] )
__children["PathFilter"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["CustomAttributes"] = GafferScene.CustomAttributes( "CustomAttributes" )
parent.addChild( __children["CustomAttributes"] )
__children["CustomAttributes"]["attributes"].addChild( Gaffer.NameValuePlug( "", Gaffer.IntPlug( "value", defaultValue = 0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member1", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["CustomAttributes"]["attributes"].addChild( Gaffer.NameValuePlug( "", Gaffer.BoolPlug( "value", defaultValue = False, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "member2", Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic ) )
__children["CustomAttributes"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Spreadsheet"] = Gaffer.Spreadsheet( "Spreadsheet" )
parent.addChild( __children["Spreadsheet"] )
__children["Spreadsheet"]["rows"].addColumn( Gaffer.NameValuePlug( "", Gaffer.BoolPlug( "value", defaultValue = False, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "sssEnabled", Gaffer.Plug.Flags.Default ), adoptEnabledPlug = True )
__children["Spreadsheet"]["rows"].addColumn( Gaffer.NameValuePlug( "", Gaffer.IntPlug( "value", defaultValue = 0, flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ), True, "sssSamples", Gaffer.Plug.Flags.Default ), adoptEnabledPlug = True )
__children["Spreadsheet"]["rows"].addRows( 2 )
__children["Spreadsheet"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SetQuery"] = GafferScene.SetQuery( "SetQuery" )
parent.addChild( __children["SetQuery"] )
__children["SetQuery"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SetFilter"] = GafferScene.SetFilter( "SetFilter" )
parent.addChild( __children["SetFilter"] )
__children["SetFilter"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Expression1"] = Gaffer.Expression( "Expression1" )
parent.addChild( __children["Expression1"] )
__children["Expression1"]["__in"].addChild( Gaffer.StringVectorDataPlug( "p0", defaultValue = IECore.StringVectorData( [  ] ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Expression1"]["__out"].addChild( Gaffer.StringPlug( "p0", direction = Gaffer.Plug.Direction.Out, defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, substitutions = IECore.StringAlgo.Substitutions.NoSubstitutions ) )
__children["Expression1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShatteredTreeBothSets"] = GafferScene.Cube( "ShatteredTreeBothSets" )
parent.addChild( __children["ShatteredTreeBothSets"] )
__children["ShatteredTreeBothSets"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SetQuery1"] = GafferScene.SetQuery( "SetQuery1" )
parent.addChild( __children["SetQuery1"] )
__children["SetQuery1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"] = GafferScene.Parent( "Parent" )
parent.addChild( __children["Parent"] )
__children["Parent"]["children"].addChild( GafferScene.ScenePlug( "child1", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"]["children"].addChild( GafferScene.ScenePlug( "child2", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Parent"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShatteredRock"] = GafferScene.Cube( "ShatteredRock" )
parent.addChild( __children["ShatteredRock"] )
__children["ShatteredRock"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["ShatteredTreeBothSets1"] = GafferScene.Cube( "ShatteredTreeBothSets1" )
parent.addChild( __children["ShatteredTreeBothSets1"] )
__children["ShatteredTreeBothSets1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Tree"]["name"].setValue( 'tree' )
__children["Tree"]["transform"]["translate"].setValue( imath.V3f( -1.52268434, 0, 0 ) )
__children["Tree"]["__uiPosition"].setValue( imath.V2f( -10.1999979, 4.05000019 ) )
__children["Group"]["in"][0].setInput( __children["Tree"]["out"] )
__children["Group"]["in"][1].setInput( __children["ShatteredTree"]["out"] )
__children["Group"]["in"][2].setInput( __children["ShatteredTreeBothSets"]["out"] )
__children["Group"]["name"].setValue( 'vegetation' )
__children["Group"]["__uiPosition"].setValue( imath.V2f( -1.29824305, -12.2781248 ) )
__children["ShatteredTree"]["name"].setValue( 'shatteredTree' )
__children["ShatteredTree"]["sets"].setValue( 'FX_BREAK' )
__children["ShatteredTree"]["__uiPosition"].setValue( imath.V2f( 7.55087757, 3.9359386 ) )
__children["Set"]["in"].setInput( __children["Parent"]["out"] )
__children["Set"]["filter"].setInput( __children["PathFilter"]["out"] )
__children["Set"]["mode"].setValue( 1 )
__children["Set"]["name"].setValue( 'VEGETATION' )
__children["Set"]["__uiPosition"].setValue( imath.V2f( 0.201756954, -43.3427773 ) )
__children["PathFilter"]["paths"].setValue( IECore.StringVectorData( [ '/vegetation' ] ) )
__children["PathFilter"]["__uiPosition"].setValue( imath.V2f( 13.2017574, -37.2601433 ) )
__children["CustomAttributes"]["in"].setInput( __children["Set"]["out"] )
__children["CustomAttributes"]["filter"].setInput( __children["SetFilter"]["out"] )
__children["CustomAttributes"]["attributes"]["member1"].setInput( __children["Spreadsheet"]["out"]["sssSamples"] )
__children["CustomAttributes"]["attributes"]["member2"].setInput( __children["Spreadsheet"]["out"]["sssEnabled"] )
__children["CustomAttributes"]["__uiPosition"].setValue( imath.V2f( 0.201756954, -61.1155434 ) )
__children["Spreadsheet"]["selector"].setInput( __children["Expression1"]["__out"]["p0"] )
__children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["name"].setValue( 'sssEnabled' )
__children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["enabled"].setValue( False )
__children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["name"].setValue( 'sssSamples' )
__children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["enabled"].setValue( False )
__children["Spreadsheet"]["rows"][1]["name"].setValue( 'FX_BREAK' )
__children["Spreadsheet"]["rows"][1]["cells"]["sssEnabled"]["value"]["name"].setValue( 'sssEnabled' )
__children["Spreadsheet"]["rows"][1]["cells"]["sssSamples"]["value"]["name"].setValue( 'sssSamples' )
__children["Spreadsheet"]["rows"][1]["cells"]["sssSamples"]["value"]["enabled"].setValue( False )
__children["Spreadsheet"]["rows"][2]["name"].setValue( 'VEGETATION' )
__children["Spreadsheet"]["rows"][2]["cells"]["sssEnabled"]["value"]["name"].setValue( 'sssEnabled' )
__children["Spreadsheet"]["rows"][2]["cells"]["sssEnabled"]["value"]["enabled"].setValue( False )
__children["Spreadsheet"]["rows"][2]["cells"]["sssSamples"]["value"]["name"].setValue( 'sssSamples' )
__children["Spreadsheet"]["rows"][2]["cells"]["sssSamples"]["value"]["value"].setValue( 4 )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"], 'ui:scene:acceptsSetName:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"], 'ui:scene:acceptsSetNames:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"], 'ui:scene:acceptsSetExpression:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"], 'spreadsheet:plugMenu:includeAsAncestor', True )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"], 'spreadsheet:plugMenu:ancestorLabel', 'Value and Switch' )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"], 'nameValuePlugPlugValueWidget:ignoreNamePlug', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["name"], 'ui:scene:acceptsSetName:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["name"], 'ui:scene:acceptsSetNames:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["name"], 'ui:scene:acceptsSetExpression:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["name"], 'ui:scene:acceptsAttributeName', True )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["value"], 'ui:scene:acceptsSetName:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["value"], 'ui:scene:acceptsSetNames:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["value"], 'ui:scene:acceptsSetExpression:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["enabled"], 'ui:scene:acceptsSetName:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["enabled"], 'ui:scene:acceptsSetNames:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssEnabled"]["value"]["enabled"], 'ui:scene:acceptsSetExpression:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"], 'ui:scene:acceptsSetName:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"], 'ui:scene:acceptsSetNames:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"], 'ui:scene:acceptsSetExpression:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"], 'spreadsheet:plugMenu:includeAsAncestor', True )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"], 'spreadsheet:plugMenu:ancestorLabel', 'Value and Switch' )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"], 'nameValuePlugPlugValueWidget:ignoreNamePlug', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["name"], 'ui:scene:acceptsSetName:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["name"], 'ui:scene:acceptsSetNames:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["name"], 'ui:scene:acceptsSetExpression:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["name"], 'ui:scene:acceptsAttributeName', True )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["value"], 'ui:scene:acceptsSetName:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["value"], 'ui:scene:acceptsSetNames:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["value"], 'ui:scene:acceptsSetExpression:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["enabled"], 'ui:scene:acceptsSetName:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["enabled"], 'ui:scene:acceptsSetNames:promotable', False )
Gaffer.Metadata.registerValue( __children["Spreadsheet"]["rows"][0]["cells"]["sssSamples"]["value"]["enabled"], 'ui:scene:acceptsSetExpression:promotable', False )
__children["Spreadsheet"]["__uiPosition"].setValue( imath.V2f( 23.638546, -60.7367134 ) )
__children["SetQuery"]["scene"].setInput( __children["Set"]["out"] )
__children["SetQuery"]["location"].setValue( '${scene:path}' )
__children["SetQuery"]["sets"].setInput( __children["Spreadsheet"]["enabledRowNames"] )
__children["SetQuery"]["__uiPosition"].setValue( imath.V2f( 24.9058628, -51.5068474 ) )
__children["SetFilter"]["setExpression"].setInput( __children["Spreadsheet"]["enabledRowNames"] )
__children["SetFilter"]["__uiPosition"].setValue( imath.V2f( 12.4888077, -51.2092705 ) )
__children["Expression1"]["__in"]["p0"].setInput( __children["SetQuery"]["matches"] )
__children["Expression1"]["__uiPosition"].setValue( imath.V2f( 28.2207794, -57.7335281 ) )
__children["ShatteredTreeBothSets"]["name"].setValue( 'shatteredTreeBothSets' )
__children["ShatteredTreeBothSets"]["sets"].setValue( 'FX_BREAK VEGETATION' )
__children["ShatteredTreeBothSets"]["transform"]["translate"].setValue( imath.V3f( 1.51659966, 0, 0 ) )
__children["ShatteredTreeBothSets"]["__uiPosition"].setValue( imath.V2f( 26.8714638, 3.92406297 ) )
Gaffer.Metadata.registerValue( __children["SetQuery1"], 'annotation:user:text', 'Debug query\n' )
Gaffer.Metadata.registerValue( __children["SetQuery1"], 'annotation:user:color', imath.Color3f( 0.150000006, 0.25999999, 0.25999999 ) )
__children["SetQuery1"]["scene"].setInput( __children["Set"]["out"] )
__children["SetQuery1"]["location"].setValue( '/shatteredTreeBothSets' )
__children["SetQuery1"]["sets"].setInput( __children["SetQuery"]["sets"] )
__children["SetQuery1"]["__uiPosition"].setValue( imath.V2f( 47.2129974, -51.5068398 ) )
__children["Parent"]["in"].setInput( __children["Group"]["out"] )
__children["Parent"]["parent"].setValue( '/' )
__children["Parent"]["children"][0].setInput( __children["ShatteredRock"]["out"] )
__children["Parent"]["children"][1].setInput( __children["ShatteredTreeBothSets1"]["out"] )
__children["Parent"]["__uiPosition"].setValue( imath.V2f( 0.201756954, -23.3065128 ) )
__children["ShatteredRock"]["name"].setValue( 'shatteredRock' )
__children["ShatteredRock"]["sets"].setValue( 'FX_BREAK' )
__children["ShatteredRock"]["transform"]["translate"].setValue( imath.V3f( 3.02014494, 0, 0 ) )
__children["ShatteredRock"]["__uiPosition"].setValue( imath.V2f( 21.1947689, -13.4036636 ) )
__children["ShatteredTreeBothSets1"]["name"].setValue( 'shatteredTreeBothSets' )
__children["ShatteredTreeBothSets1"]["sets"].setValue( 'FX_BREAK VEGETATION' )
__children["ShatteredTreeBothSets1"]["transform"]["translate"].setValue( imath.V3f( 4.50511885, 0, 0 ) )
__children["ShatteredTreeBothSets1"]["__uiPosition"].setValue( imath.V2f( 44.8940392, -13.3602772 ) )
__children["Expression1"]["__engine"].setValue( 'python' )
__children["Expression1"]["__expression"].setValue( 'matches = parent["__in"]["p0"]\nparent["__out"]["p0"] = matches[-1] if len( matches ) else ""' )


del __children

```

</details>

> I think it's necessary for the SetSwitch. The internal Replace node will be replacing from the ancestor match down (so descendant matches are irrelevant), so SetQuery needs to tell you which set matched that ancestor, to drive the NameSwitch that chooses the replacement.

I'm now not 100% sure this is actually true for the hypothetical Gaffer SetSwitch. I don't think Replace will care about descendant matches because it will be throwing away the entire branch below the highest match in the hierarchy. Where I still think it does matter is as an internal component for CsSetSwitch, which would just naively do the set query at every input location, and need the result to be consistent down the hierarchy. Maybe we shouldn't drive our design with this short-term consideration though? I'll check in with Alex and then we can discuss in today's catch up.

Reply:
I checked with Alex and he's OK with ignoring the CsSetSwitch use case (I've provided the original code to him separately so they can use it if needed), so I've pushed an update to strictly follow the query order regardless of inheritance. Only the [last commit](https://github.com/GafferHQ/gaffer/pull/4959/commits/58ea01b8563ed553064d0dd90862f654b5c98247) is new, the others are just the originals rebased to fix merge conflicts.

Testing with the example above, and replacing the expression with a simple `firstMatch` connection, all cases including `/shatteredTreeBothSets` now act as we want. Thanks for pushing on this one.